components/brave_rewards/browser/diagnostic_log.cc

```diff
@@ -0,0 +1,365 @@
+
+  if (!Open(file_path, file)) {
+    return false;
+  }
+
+  return true;
+}
+
+int64_t SeekFromEnd(base::File* file, const int num_lines) {
+  DCHECK(file);
+
+  if (!file->IsValid()) {
+    return 0;
+  }
+
+  if (num_lines == 0) {
+    return 0;
+  }
+
+  int64_t length = file->GetLength();
+  if (length == -1) {
+    return -1;
+  }
+
+  if (length == 0) {
+    return 0;
+  }
+
+  int line_count = 0;
+
+  char chunk[kChunkSize];
+  int64_t chunk_size = kChunkSize;
+  int64_t last_chunk_size = 0;
+
+  if (file->Seek(base::File::FROM_END, 0) == -1) {
+    return -1;
+  }
+
+  do {
+    if (chunk_size > length) {
+      chunk_size = length;
+    }
+
+    const int64_t seek_offset = chunk_size + last_chunk_size;
+    if (file->Seek(base::File::FROM_CURRENT, -seek_offset) == -1) {
+      return -1;
+    }
+
+    if (file->ReadAtCurrentPos(chunk, chunk_size) == -1) {
+      return -1;
+    }
+
+    for (int i = chunk_size - 1; i >= 0; i--) {
+      if (chunk[i] == '\n') {
+        line_count++;
+        if (line_count == num_lines + 1) {
+          return length;
+        }
+      }
+
+      length--;
+    }
+
+    last_chunk_size = chunk_size;
+  } while (length > 0);
+
+  return length;
+}
+
+bool TrimBeginningOfFile(base::File* file,
+                         int keep_num_lines) {
+  DCHECK(file);
+
+  if (file->GetLength() == 0) {
+    return true;
+  }
+
+  const int64_t offset = SeekFromEnd(file, keep_num_lines);
+  if (offset == -1) {
+    return false;
+  }
+
+  if (offset == 0) {
+    return true;
+  }
+
+  if (file->Seek(base::File::FROM_BEGIN, offset) == -1) {
+    return false;
+  }
+
+  const int64_t size = file->GetLength() - offset;
+  std::unique_ptr<char[]> buffer = std::make_unique<char[]>(size + 1);
+
+  if (file->ReadAtCurrentPos(buffer.get(), size) == -1) {
+    return false;
+  }
+
+  if (file->Seek(base::File::FROM_BEGIN, 0) == -1) {
+    return false;
+  }
+
+  const std::string data = std::string(buffer.get());
+  const int64_t new_size = data.size();
+
+  if (file->WriteAtCurrentPos(data.c_str(), new_size) == -1) {
+    return false;
+  }
+
+  if (!file->SetLength(new_size)) {
+    return false;
+  }
+
+  return true;
+}
+
+bool MaybeTrimBeginningOfFile(base::File* file,
+                              const int64_t max_file_size,
+                              const int keep_num_lines,
+                              const bool first_write) {
+  DCHECK(file);
+
+  const int64_t length = file->GetLength();
+  if (length == -1) {
+    return false;
+  }
+
+  // We do not trim the log on first run so that if the browser
+  // crashes and we investigate the log with the user they are able to
+  // re-run the browser without losing past logs.
+  if (!first_write && length <= max_file_size) {
+    return true;
+  }
+
+  if (!TrimBeginningOfFile(file, keep_num_lines)) {
+    return false;
+  }
+
+  return true;
+}
+
+std::string ReadLastNLinesOnFileTaskRunner(
+    const base::FilePath& file_path,
+    const int num_lines) {
```

Comment:
For this and other similar value-type params, can we use [this rule](https://google.github.io/styleguide/cppguide.html#Use_of_const)?

> For a function parameter passed by value, const has no effect on the caller, thus is not recommended in function declarations.

Reply:
Yes, will do. (I like that rule too, I just kept these as they were when copying.)